### PR TITLE
SNI based Configuration for traffic server

### PR DESCRIPTION
**Pull request by persiaAziz**

### CODE REVIEW

File: iocore/net/LuaSNIConfig.cc [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                                                               {{"NONE", 0}, {"MODERATE", 1}, {"STRICT", 2}}};
+TsConfigDescriptor LuaSNIConfig::Item::TUNNEL_DEST_DESCRIPTOR = {TsConfigDescriptor::Type::STRING, "String", TS_tunnel_route,
+                                                                 "tunnel route destination"};
+TsConfigDescriptor LuaSNIConfig::Item::CLIENT_CERT_DESCRIPTOR = {TsConfigDescriptor::Type::STRING, "String", TS_client_cert,
+                                                                 "Client certificate to present to the next hop server"};
+TsConfigDescriptor LuaSNIConfig::Item::VERIFY_NEXT_SERVER_DESCRIPTOR = {TsConfigDescriptor::Type::INT, "Int",
+                                                                        TS_verify_origin_server, "Next hop verification level"};
+
+ts::Errata
+LuaSNIConfig::loader(lua_State *L)
```
Comment (jpeach) [inline iocore/net/LuaSNIConfig.cc]: [BindingInstance](https://github.com/apache/trafficserver/blob/master/lib/bindings/bindings.h) is the canonical way to make Lua APIs. Please use that instead.

File: iocore/net/Makefile.am [hunk 2]
```diff
@@ -17,9 +17,11 @@
 #  limitations under the License.
 
 AM_CPPFLAGS += \
+  $(LUAJIT_CPPFLAGS) \
   $(iocore_include_dirs) \
   -I$(abs_top_srcdir)/lib \
   -I$(abs_top_srcdir)/lib/records \
+  -I$(abs_top_srcdir)/lib/tsconfig \
```
Comment (SolidWallOfCode) [inline iocore/net/Makefile.am]: Don't do this - just `#include <tsconfig/Errata.h>`.

File: iocore/net/LuaSNIConfig.cc (hunk 1 above)
Comment (SolidWallOfCode) [reply iocore/net/LuaSNIConfig.cc]: This is based on the TsLuaConfig work which I did a brief presentation on two summits back, and a longer at the most recent summit. I've discussed it at length with Kit as well and he's OK with this approach. This is really a prototype of the work, to get this PR running.

File: iocore/net/LuaSNIConfig.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+#include <vector>
+
+using ts::Errata;
+#define LUA_ENUM(L, name, val)                  \
+  lua_pushlstring(L, #name, sizeof(#name) - 1); \
+  lua_pushnumber(L, val);                       \
+  lua_settable(L, -3);
+
+constexpr char TS_fqdn[]                 = "fqdn";
```
Comment (SolidWallOfCode) [inline iocore/net/LuaSNIConfig.h]: Should probably make these `string_view` instances.

File: iocore/net/LuaSNIConfig.cc (hunk 1 above)
Comment (jpeach) [reply iocore/net/LuaSNIConfig.cc]: I really don't see why we need another way to bind Lua APIs.

File: proxy/config/ssl_server_name.config.default [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ Example:
+
+
+params for client verification
+        NONE, MODERATE, STRICT
+
+sni_config = {
+                {
+                    fqdn='one.com',
+                    disable_h2 = <TRUE|FALSE>,
```
Comment (masaori335) [inline proxy/config/ssl_server_name.config.default]: It'd be great if this is protocol list. e.g. `proto={h1}`, `proto={h1,h2}`.
Otherwise, when new protocol is defined, we need to add another flag for it.

File: proxy/config/ssl_server_name.config.default [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+   tunnel_route  - sets the e2e tunnel route
+ Example:
+
+
+params for client verification
+        NONE, MODERATE, STRICT
+
+sni_config = {
+                {
+                    fqdn='one.com',
```
Comment (masaori335) [inline proxy/config/ssl_server_name.config.default]: Is this replace `ssl_multicert.config` ? If so, is this support default cert like `dest_ip=*` in `ssl_multicert.config` ?

File: iocore/net/LuaSNIConfig.cc (hunk 1 above)
Comment (SolidWallOfCode) [reply iocore/net/LuaSNIConfig.cc]: https://cwiki.apache.org/confluence/display/TS/Presentations+-+2017?preview=/70255385/74684802/TsLuaConfig-Fall-2017.pdf

**Comment by SolidWallOfCode** [comment]
[approve ci]

File: cmd/traffic_layout/traffic_layout.cc [hunk 6]
```diff
@@ -172,6 +172,7 @@ produce_layout(bool json)
   print_var("remap.config", RecConfigReadConfigPath("proxy.config.url_remap.filename"), json);
   print_var("plugin.config", RecConfigReadConfigPath(nullptr, "plugin.config"), json);
   print_var("ssl_multicert.config", RecConfigReadConfigPath("proxy.config.ssl.server.multicert.filename"), json);
+  print_var("ssl_server_name.config", RecConfigReadConfigPath("proxy.config.ssl.servername.filename"), json);
```
Comment (SolidWallOfCode) [inline cmd/traffic_layout/traffic_layout.cc]: Why isn't this `SSL_SERVER_NAME_CONFIG`?

File: iocore/net/Makefile.am [hunk 7]
```diff
@@ -17,6 +17,7 @@
 #  limitations under the License.
 
 AM_CPPFLAGS += \
+  $(LUAJIT_CPPFLAGS) \
```
Comment (SolidWallOfCode) [inline iocore/net/Makefile.am]: Why are these needed?

File: iocore/net/P_SNIActionPerformer.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
+/** @file
+
+  A brief file description
```
Comment (SolidWallOfCode) [inline iocore/net/P_SNIActionPerformer.h]: Less brief, please.

File: iocore/net/P_SNIActionPerformer.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+extern Map<int, SSLNextProtocolSet *> snpsMap;
+// enum of all the actions
+enum AllActions {
+  TS_DISABLE_H2 = 0,
+  TS_VERIFY_CLIENT, // this applies to server side vc only
+  TS_TUNNEL_ROUTE,  // blind tunnel action
+};
+
+/** action for setting next hop properties should be listed in the following enum*/
+enum PropertyActions { TS_VERIFY_SERVER = 200, TS_CLIENT_CERT };
```
Comment (SolidWallOfCode) [inline iocore/net/P_SNIActionPerformer.h]: Why `200`?

File: iocore/net/SSLNetVConnection.cc [hunk 10]
```diff
@@ -31,9 +31,13 @@
 #include "BIO_fastopen.h"
 #include "Log.h"
 #include "P_SSLClientUtils.h"
+#include "P_SSLSNI.h"
+#include "HttpTunnel.h"
 
+#include <unordered_set>
 #include <climits>
 #include <string>
+#include <stdbool.h>
```
Comment (SolidWallOfCode) [inline iocore/net/SSLNetVConnection.cc]: What is this for?

File: iocore/net/SSLNetVConnection.cc [hunk 11]
```diff
@@ -31,9 +31,13 @@
 #include "BIO_fastopen.h"
 #include "Log.h"
 #include "P_SSLClientUtils.h"
+#include "P_SSLSNI.h"
+#include "HttpTunnel.h"
 
+#include <unordered_set>
```
Comment (SolidWallOfCode) [inline iocore/net/SSLNetVConnection.cc]: This will raise strong objections - can you try to use `HashMap` from `lib/ts/Map.h`?

File: iocore/net/SSLUtils.cc [hunk 12]
```diff
@@ -407,10 +414,16 @@ ssl_cert_callback(SSL *ssl, void * /*arg*/)
 /*
  * Cannot stop this callback. Always reeneabled
  */
+extern SNIActionPerformer sni_action_performer;
 static int
 ssl_servername_only_callback(SSL *ssl, int * /* ad */, void * /*arg*/)
 {
-  SSLNetVConnection *netvc = SSLNetVCAccess(ssl);
+  SSLNetVConnection *netvc = (SSLNetVConnection *)SSL_get_app_data(ssl);
```
Comment (SolidWallOfCode) [inline iocore/net/SSLUtils.cc]: No C-style casts.

File: lib/ts/Vec.h [hunk 13]
```diff
@@ -853,9 +853,8 @@ template <class C, class A, int S>
 inline void
 Vec<C, A, S>::free_and_clear()
 {
-  for (int x = 0; x < n; x++)
-    if (v[x])
-      A::free(v[x]);
+  for (size_t x = 0; x < (n); x++)
```
Comment (SolidWallOfCode) [inline lib/ts/Vec.h]: This is going to be problematic since `Vec.h` has been removed. You will likely need to convert use of `Vec` to `std::vector`.

File: proxy/config/ssl_server_name.config.default (hunk 4 above)
Comment (SolidWallOfCode) [reply proxy/config/ssl_server_name.config.default]: Yes, I think we'll clean that up when we get the TsLuaConfig more robust.

File: proxy/http/HttpSM.cc [hunk 14]
```diff
@@ -69,6 +70,7 @@
 #define USE_NEW_EMPTY_MIOBUFFER
 
 extern int cache_config_read_while_writer;
+extern std::unordered_map<std::string, std::string> TunnelMap; // stores the name of the servers to tunnel to
```
Comment (SolidWallOfCode) [inline proxy/http/HttpSM.cc]: It seems wrong to have this in so many places - is there not a flag that could be set earlier in the processing to avoid repeated lookups?

File: proxy/http/HttpSM.cc [hunk 15]
```diff
@@ -533,8 +535,43 @@ HttpSM::setup_client_read_request_header()
...
+      t_state.hdr_info.client_request.create(HTTP_TYPE_REQUEST);
+      t_state.hdr_info.client_request.method_set(HTTP_METHOD_CONNECT, HTTP_LEN_CONNECT);
+      t_state.hdr_info.client_request.url_create(&u);
+      u.scheme_set(URL_SCHEME_TUNNEL, URL_LEN_TUNNEL);
+      t_state.hdr_info.client_request.url_set(&u);
+      std::unordered_map<std::string, std::string>::const_iterator got = TunnelMap.find(std::string(ssl_vc->serverName));
+
+      if (got != TunnelMap.end()) {
+        ts::string_view addr, port;
+        if (ats_ip_parse(ts::string_view(got->second), &addr, &port) == 0) {
```
Comment (SolidWallOfCode) [inline proxy/http/HttpSM.cc]: This should have been parsed at configuration load time, not every transaction.

File: proxy/http/HttpSM.cc [hunk 16]
```diff
@@ -1335,6 +1372,42 @@ HttpSM::state_api_callout(int event, void *data)
...
   case EVENT_NONE:
+    if (cur_hook_id == TS_HTTP_TXN_START_HOOK && t_state.client_info.port_attribute == HttpProxyPort::TRANSPORT_BLIND_TUNNEL) {
+      /* Creating the request object early to set the host header and port for blind tunneling here for the
+plugins required to work with sni_routing.
+*/
+      // Plugins triggered on txn_start_hook will get the host and port at that point
+      // We've received a request on a port which we blind forward
+      URL u;
+
+      t_state.hdr_info.client_request.create(HTTP_TYPE_REQUEST);
```
Comment (SolidWallOfCode) [inline proxy/http/HttpSM.cc]: Isn't this `setup_blind_tunnel_port` again? This should be consolidated, not cut and pasted.

File: proxy/http/HttpTransact.cc [hunk 17]
```diff
@@ -542,34 +542,18 @@ HttpTransact::Forbidden(State *s)
 void
 HttpTransact::HandleBlindTunnel(State *s)
 {
-  NetVConnection *vc         = s->state_machine->ua_session->get_netvc();
-  bool inbound_transparent_p = vc->get_is_transparent();
```
Comment (SolidWallOfCode) [inline proxy/http/HttpTransact.cc]: There is specific handling here for transparent connections, but I don't see that in the replacement code. Is it no longer needed? If so, why not?

Comment (persiaAziz) [reply proxy/http/HttpTransact.cc]: This flag never had any use. Line 596: `if (!(inbound_transparent_p || url_remap_success) || remap_redirect != nullptr)` , url_remap_process is always true , which makes this flag pointless. Again remap_redirect is set to nullptr in 587. Hence the whole if condition is always false.

File: proxy/http/HttpSM.cc (hunk 16 above)
Comment (persiaAziz) [reply proxy/http/HttpSM.cc]: well , not quite. `if (!t_state.hdr_info.client_request.url_get()->host_get(&host_len))` this checking in `setup_blind_tunnel_port` is required to avoid setting the hosts again in case of plugin interference at the start hook.

File: proxy/hdrs/URL.cc [hunk 18]
```diff
@@ -1206,6 +1206,7 @@ ParseResult
 url_parse_no_path_component_breakdown(HdrHeap *heap, URLImpl *url, const char **start, const char *end, bool copy_strings_p)
 {
   ParseResult zret = url_parse_scheme(heap, url, start, end, copy_strings_p);
+  url->m_port      = url_canonicalize_port(url->m_url_type, url->m_port);
```
Comment (persiaAziz) [inline proxy/hdrs/URL.cc]: looks like , I can not set the ports here since it breaks people's test :/

File: proxy/http/HttpSM.cc (hunk 15 above)
Comment (persiaAziz) [reply proxy/http/HttpSM.cc]: sounds good

File: proxy/http/HttpSM.cc (hunk 14 above)
Comment (persiaAziz) [reply proxy/http/HttpSM.cc]: well the lookup happens twice. First in the servername callback. second in the host setup code. the callback depends on the version of openssl and is not guaranteed to get triggered (in case CONNECT for e.g.,).

File: proxy/config/ssl_server_name.config.default (hunk 5 above)
Comment (persiaAziz) [reply proxy/config/ssl_server_name.config.default]: yes it does support wild cards. Replacing ssl_multiert.config is my future plan. But right now I guess replacing ssl_multicert will be a big change for a lot of users

File: cmd/traffic_layout/traffic_layout.cc (hunk 6 above)
Comment (persiaAziz) [reply cmd/traffic_layout/traffic_layout.cc]: I was matching with the other print_vars -_-

File: lib/ts/Vec.h (hunk 13 above)
Comment (persiaAziz) [reply lib/ts/Vec.h]: yah..  :-<

File: iocore/net/P_SNIActionPerformer.h (hunk 8 above)
Comment (persiaAziz) [reply iocore/net/P_SNIActionPerformer.h]: I like netbeans

File: iocore/net/P_SNIActionPerformer.h (hunk 9 above)
Comment (persiaAziz) [reply iocore/net/P_SNIActionPerformer.h]: random number to keep property actions far from the SNI based actions.

File: iocore/net/SSLNetVConnection.cc (hunk 11 above)
Comment (persiaAziz) [reply iocore/net/SSLNetVConnection.cc]: alright. I ported this from my 5.3 patch. Forgot to change this

File: proxy/http/HttpSM.cc (hunk 14 above)
Comment (persiaAziz) [reply proxy/http/HttpSM.cc]: but then again , it does look for the servername I am setting in sslvc. Let me think about it more

File: proxy/http/HttpSM.cc (hunk 16 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpSM.cc]: Weak. This code is extremely similar, it will need to be cleaned up later. Look at 1390-1401 vs. 553-563. And 1379-1383 vs. 546-550.

File: iocore/net/LuaSNIConfig.cc (hunk 1 above)
Comment (SolidWallOfCode) [reply iocore/net/LuaSNIConfig.cc]: @jpeach - I think the schema based approach is better and it's reasonable to try it while we are still getting set up for using Lua configs.  There's an example of using LuaBindings, I want to try an example using TsLuaConfig.

**Comment by persiaAziz** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[ci approve autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
Why does the HttpSM include unordered_map ??

**Comment by persiaAziz** [comment]
getting rid of the includes, unordered_map is not really doing anything

**Comment by persiaAziz** [comment]
fixing the Makefile of traffic_wccp.
